File: backend/app/schemas/customer.py
from datetime import datetime
from typing import List, Optional

from pydantic import Field, field_validator

from app.schemas.common import ApiSchema

class CustomerCreate(ApiSchema):
    name: str = Field(..., min_length=1, max_length=100)
    contact_name: str = Field(..., min_length=1, max_length=50)
    contact_phone: str = Field(..., min_length=1, max_length=20)
    level_id: str
    address: Optional[str] = None
    remark: Optional[str] = None
    image_urls: Optional[List[str]] = None


class CustomerUpdate(ApiSchema):
    name: Optional[str] = Field(None, max_length=100)
    contact_name: Optional[str] = Field(None, max_length=50)
    contact_phone: Optional[str] = None
    level_id: Optional[str] = None
    address: Optional[str] = None
    remark: Optional[str] = None
    image_urls: Optional[List[str]] = None


class CustomerOut(ApiSchema):
    id: str
    name: str
    contact_name: str
    contact_phone: str
    level_id: str
    level_name: Optional[str] = None
    address: Optional[str]
    remark: Optional[str]
    image_urls: Optional[List[str]]
    total_spent: float = 0
    order_count: int = 0
    last_order_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("id", "level_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v)


class CustomerLevelCreate(ApiSchema):
    name: str = Field(..., min_length=1, max_length=50)
    min_spent: float = Field(0, ge=0)
    sort_order: int = Field(0)
    is_default: bool = Field(False)


class CustomerLevelUpdate(ApiSchema):
    name: Optional[str] = Field(None, max_length=50)
    min_spent: Optional[float] = Field(None, ge=0)
    sort_order: Optional[int] = None
    is_default: Optional[bool] = None


class CustomerLevelOut(ApiSchema):
    id: str
    name: str
    min_spent: float
    sort_order: int
    is_default: bool
    created_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v)


class MemberPriceCreate(ApiSchema):
    product_id: str
    level_id: str
    price: float = Field(..., ge=0)


class MemberPriceUpdate(ApiSchema):
    price: float = Field(..., ge=0)


class MemberPriceOut(ApiSchema):
    id: str
    product_id: str
    level_id: str
    product_name: Optional[str] = None
    barcode: Optional[str] = None
    level_name: Optional[str] = None
    price: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("id", "product_id", "level_id", mode="before")
    @classmethod
    def uuid_to_str(cls, v):
        return str(v)

File: frontend/src/services/customer.ts
import { request } from '@umijs/max';
import { collectResponsePages } from './pagination';
export interface MemberPriceItem {
  id: string;
  product_id: string;
  level_id: string;
  barcode?: string;
  product_name?: string;
  level_name?: string;
  price: number;
}
export interface CustomerOut {
  id: string;
  name: string;
  contact_name: string;
  contact_phone: string;
  level_id: string;
  level_name: string | null;
  address: string | null;
  remark: string | null;
  image_urls: string[] | null;
  total_spent: number;
  order_count: number;
  last_order_at: string | null;
  created_at: string;
  updated_at: string;
}
export async function listCustomers(params?: { keyword?: string; page?: number; page_size?: number }) {
  return request<API.ResponseBase<API.PaginatedData>>('/api/v1/customers', { method: 'GET', params });
}
export async function listAllCustomers() {
  return collectResponsePages((page, pageSize) =>
    listCustomers({ page, page_size: pageSize }),
  );
}
export async function getCustomer(id: string) {
  return request<API.ResponseBase<CustomerOut>>(`/api/v1/customers/${id}`, { method: 'GET' });
}
export async function createCustomer(data: { name: string; contact_name: string; contact_phone: string; address?: string; level_id?: string; remark?: string; image_urls?: string[] }) {
  return request<API.ResponseBase>('/api/v1/customers', { method: 'POST', data });
}
export async function updateCustomer(id: string, data: { name?: string; contact_name?: string; contact_phone?: string; address?: string; level_id?: string; remark?: string; image_urls?: string[] }) {
  return request<API.ResponseBase>(`/api/v1/customers/${id}`, { method: 'PUT', data });
}
export async function listLevels(params?: { page?: number; page_size?: number }) {
  return request<API.ResponseBase<API.PaginatedData>>('/api/v1/levels', { method: 'GET', params });
}
export async function listAllLevels() {
  return collectResponsePages((page, pageSize) =>
    listLevels({ page, page_size: pageSize }),
  );
}
export async function createLevel(data: { name: string; min_spent: number; sort_order?: number; is_default?: boolean }) {
  return request<API.ResponseBase>('/api/v1/levels', { method: 'POST', data });
}
export async function updateLevel(id: string, data: { name?: string; min_spent?: number; sort_order?: number; is_default?: boolean }) {
  return request<API.ResponseBase>(`/api/v1/levels/${id}`, { method: 'PUT', data });
}
export async function deleteLevel(id: string) {
  return request<API.ResponseBase>(`/api/v1/levels/${id}`, { method: 'DELETE' });
}
